clients/wallet/src/horizon.rs

```diff
@@ -426,31 +442,74 @@ impl HorizonClient for reqwest::Client {
 		&self,
 		transaction_envelope: TransactionEnvelope,
 		is_public_network: bool,
+		max_retries: u8,
+		max_backoff_delay: u8,
 	) -> Result<TransactionResponse, Error> {
+		let seq_no = transaction_envelope.sequence_number();
 		let transaction_xdr = transaction_envelope.to_base64_xdr();
 		let transaction_xdr = std::str::from_utf8(&transaction_xdr).map_err(Error::Utf8Error)?;
 
-		let base_url = horizon_url(is_public_network);
-		let url = format!("{}/transactions", base_url);
-
 		let params = [("tx", &transaction_xdr)];
-		let response =
-			self.post(url).form(&params).send().await.map_err(Error::HttpFetchingError)?;
-
-		interpret_response::<TransactionResponse>(response).await.map_err(|e| match e {
-			Error::HorizonSubmissionError { title, status, reason, envelope_xdr }
-				if envelope_xdr.is_none() =>
-			{
-				// let's add a transaction envelope, if possible
-				Error::HorizonSubmissionError {
-					title,
-					status,
-					reason,
-					envelope_xdr: Some(transaction_xdr.to_string()),
-				}
-			},
-			other => other,
-		})
+
+		let mut server_error_count = 0;
+		let mut exponent_counter = 1;
+
+		loop {
+			let need_fallback = if server_error_count == max_retries {
+				// reset in case the fallback fails but original url could be running again.
+				server_error_count = 0;
+				true
+			} else {
+				false
+			};
+
+			let base_url = horizon_url(is_public_network, need_fallback);
+			let url = format!("{}/transactions", base_url);
+
+			let response =
+				ready(self.post(url).form(&params).send().await.map_err(Error::HttpFetchingError))
+					.and_then(|response| async move {
+						interpret_response::<TransactionResponse>(response).await
+					})
+					.await;
+
+			match response {
+				Err(e) if e.is_recoverable() || e.is_server_error() => {
+					if e.is_server_error() {
+						server_error_count += 1;
+					}
+
+					// let's wait awhile before resubmitting.
+					tracing::warn!(
+						"submission failed for transaction with sequence number {seq_no:?}: {e:?}"
+					);
+					// exponentially sleep before retrying again
+					sleep(Duration::from_secs(2u64.pow(exponent_counter))).await;
+					tracing::debug!("resubmitting transaction with sequence number {seq_no:?}...");
+
+					// retry/resubmit again
+					if exponent_counter > u32::from(max_backoff_delay) {
+						exponent_counter = 1;
```

Comment:
Instead of re-setting the increment and starting from `1` again, as a caller of this function I would naturally assume that once it reaches the maximum, it just stays at the maximum (and does not flip back to a lower delay again). So I would suggest instead of the `if-else` we just use an `if` statement where we check if the exponent counter is less than the limit and increment it if that's the case. And we do nothing in the other case.

Reply:
Also, the naming is a bit confusing. We have the `max_backoff_delay` here, which without knowing the implementation would make you think that this is some value in seconds, which is the maxium backoff delay, meaning that the transaction submission will not be delayed more than this value. _But_ what this actually is, is an exponent for the exponential backoff calculation. So we'd at least have to rename it so we don't cause confusion about it.

However, I think I'd prefer if we rename the variable to `max_backoff_delay_seconds`. Then in this check here, we'd do (also including the changes from my previous comment):
```
if 2u64pow(exponent_counter) < max_backoff_delay_seconds {
	exponent_counter += 1;
}
```


And for the sleep statement above, we'd do:
```
let sleep_duration = std::cmp::min(2u64.pow(exponent_counter, max_backoff_delay_seconds)));
sleep(sleep_duration);
```


WDYT @b-yap?

Reply:
done